### PR TITLE
Delete atomic and durable

### PR DESCRIPTION
**Pull request by Unrud**
See #440

### CODE REVIEW

**Comment by liZe** [comment]
Looks good for me (at least for posix systems), but I've no experience at all in atomic writes.

@untitaker @kentonv anything to say about this?

File: radicale/storage.py [hunk 1]
```diff
@@ -505,13 +505,38 @@ def update(self, href, vobject_item, etag=None):
...
+                try:
+                    os.rmdir(self._filesystem_path)
+                except OSError:
+                    while True:
+                        tmp_filesystem_path = os.path.join(
+                            os.path.dirname(self._filesystem_path),
+                            ".Radicale.tmp-" + hex(getrandbits(32))[2:])
+                        if not os.path.exists(tmp_filesystem_path):
+                            break
+                    os.rename(self._filesystem_path, tmp_filesystem_path)
```
Comment (untitaker) [inline radicale/storage.py]: Couldn't we just use the stdlib tmpdir functions for this? I.e. create a tmpdir, move the collection into that folder and delete the tmpdir. Less performant but also less code to overlook.

Also there'd be a race condition here if there was no global lock.

File: radicale/storage.py [hunk 2]
```diff
@@ -505,13 +505,38 @@ def update(self, href, vobject_item, etag=None):
     def delete(self, href=None, etag=None):
         # TODO: use etag in code and test it here
         # TODO: use returned object in code
+        def sync_directory(path):
```
Comment (untitaker) [inline radicale/storage.py]: I feel like this would be better placed in another module within Radicale.

Maybe I should expose this in atomicwrites?

**Comment by untitaker** [comment]
Unrud's code is always solid, so I only left stylistic nitpicks.

File: radicale/storage.py (hunk 2 above)
Comment (liZe) [inline radicale/storage.py]: > Maybe I should expose this in atomicwrites?

It makes sense.

Comment (Unrud) [inline radicale/storage.py]: Good idea.

**Comment by Unrud** [comment]
> Looks good for me (at least for posix systems),

I can't find any documentation for Windows on this issue. I don't know how to sync meta data or flush the on-disk cache. Replace operations don't seem to be atomic. Moving a file seems to be atomic, as long as the target doesn't already exist and it's within the same volume.
Windows has Transactional NTFS which would be perfect for this, but it's deprecated.

**Comment by Unrud** [comment]
Rebased. I think this can be merged now. If the ``sync_directory`` function gets exposed by ``atomicwrites`` we can later switch to it.

**Comment by liZe** [comment]
Thank you!

**Pull request merged**